src/multivariate/dirichlet.jl

```diff
@@ -375,3 +375,63 @@ function fit_mle(::Type{<:Dirichlet}, P::AbstractMatrix{Float64},
     elogp = mean_logp(suffstats(Dirichlet, P, w))
     fit_dirichlet!(elogp, α; maxiter=maxiter, tol=tol, debug=debug)
 end
+
+## Differentiation
+function ChainRulesCore.frule((_, Δalpha)::Tuple{Any,Any}, DT::Union{Type{Dirichlet{T}}, Type{Dirichlet}}, alpha::AbstractVector{T}; check_args = true) where {T}
+    d = DT(alpha; check_args=check_args)
+    Δalpha = ChainRulesCore.unthunk(Δalpha)
+    ∂alpha0 = sum(Δalpha)
+    ∂lmnB::typeof(∂alpha0) = sum(Δalpha[i] * (SpecialFunctions.digamma(alpha[i]) - SpecialFunctions.digamma(d.alpha0)) for i in eachindex(alpha))
+    backing = (alpha=Δalpha, alpha0=∂alpha0, lmnB=∂lmnB)
+    t = ChainRulesCore.Tangent{typeof(d), NamedTuple{(:alpha, :alpha0, :lmnB), Tuple{typeof(alpha), typeof(d.alpha0), typeof(d.lmnB)}}}(backing)
+    return d, t
+end
+
+function ChainRulesCore.rrule(DT::Union{Type{Dirichlet{T}}, Type{Dirichlet}}, alpha::AbstractVector{T}; check_args = true) where {T}
+    d = DT(alpha; check_args=check_args)
+    function dirichlet_pullback(d_dir)
+        d_dir = ChainRulesCore.unthunk(d_dir)
+        ∂l = d_dir.lmnB * (SpecialFunctions.digamma.(alpha) .- SpecialFunctions.digamma.(d.alpha0))
+        return (ChainRulesCore.NoTangent(), d_dir.alpha .+ d_dir.alpha0 .+ ∂l)
+    end
+    return d, dirichlet_pullback
+end
+
+function ChainRulesCore.frule((_, Δd, Δx), ::typeof(_logpdf), d::Dirichlet, x::AbstractVector{T}) where {T}
+    lp = _logpdf(d, x)
+    if !insupport(d, x)
+        return (lp, zero(lp))
+    end
+    ∂α = sum(Δd.alpha[i] * log(x[i]) for i in eachindex(x))
+    ∂l = - Δd.lmnB
+    ∂x = sum((d.alpha[i] - 1) * Δx[i] / x[i] for i in eachindex(x))
+    return (lp, ∂α + ∂l + ∂x)
```

Comment:
It seems one can combine the two loops. Again, indices might be different. One could use `zip` or pairwise summation with
```suggestion
    dlp = sum(Broadcast.instantiate(Broadcast.broadcasted(d.alpha, Δd.alpha, x, Δx) do alpha_i, Δalpha_i, x_i, Δx_i
        Δalpha_i * log(x_i) + (alpha_i - 1) * Δx_i / x_i
    end)) - Δd.lmnB
    return (lp, dlp)
```

Additionally, it seems the case `x[i] = 0` is problematic and maybe should be handled. `xlogy` could be useful for the first term, not sure about the division in the second one.

Reply:
done in a single sum using the do notation, I think a broadcast would have allocated right?

Reply:
No, in contrast to regular `broadacst` `Broadcast.instantiate(Broadcast.broadcasted(...))` doesn't allocate. It's a "trick" that is used e.g. in StatsBase but also Distributions to use pairwise summation without allocating intermediate arrays. E.g., compare
```julia
julia> using BenchmarkTools, LinearAlgebra

julia> A = vcat(1f0, fill(1f-8, 10^8));

julia> w = ones(Float32, length(A));

julia> f_dot(A, w) = dot(A, w)
f_dot (generic function with 1 method)

julia> f_sum_zip(A, w) = sum(Ai * wi for (Ai, wi) in zip(A, w))
f_sum_zip (generic function with 1 method)

julia> f_sum_broadcasted(A, w) = sum(Broadcast.instantiate(Broadcast.broadcasted(*, A, w)))
f_sum_broadcasted (generic function with 1 method)

julia> f_dot(A, w)
1.9625133f0

julia> f_sum_zip(A, w)
1.0f0

julia> f_sum_broadcasted(A, w)
1.9999989f0

julia> @btime f_dot($A, $w);
  42.887 ms (0 allocations: 0 bytes)

julia> @btime f_sum_zip($A, $w);
  140.985 ms (0 allocations: 0 bytes)

julia> @btime f_sum_broadcasted($A, $w);
  46.262 ms (0 allocations: 0 bytes)
```